### PR TITLE
[dif_sram_ctrl] Bring SRAM Controller DIF to S2

### PR DESCRIPTION
**Pull request by silvestrst**
This PR adds missing SRAM Controller implementation, and bumps the checklist to S2.

### CODE REVIEW

File: sw/device/lib/dif/dif_sram_ctrl.c [hunk 1]
```diff
@@ -64,6 +64,43 @@ static uint32_t sram_ctrl_get_status(const dif_sram_ctrl_t *sram_ctrl) {
...
+  // Scramble memory addresses.
+  uint32_t reg =
+      bitfield_bit32_write(0, SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT, true);
+  mmio_region_write32(sram_ctrl->base_addr, SRAM_CTRL_CTRL_REG_OFFSET, reg);
+
+  // Wait for address scrambling to finish.
+  dif_sram_ctrl_status_bitfield_t status;
+  do {
+    status = sram_ctrl_get_status(sram_ctrl);
+  } while (!(status & kDifSramCtrlStatusScrKeyValid));
```
Comment (mcy) [inline sw/device/lib/dif/dif_sram_ctrl.c:85]: Can you make this == 0 instead of replying on contextual conversion to bool?

Comment (silvestrst) [reply sw/device/lib/dif/dif_sram_ctrl.c]: Done.

File: sw/device/lib/dif/dif_sram_ctrl.c [hunk 2]
```diff
@@ -64,6 +64,43 @@ static uint32_t sram_ctrl_get_status(const dif_sram_ctrl_t *sram_ctrl) {
...
+  reg = bitfield_bit32_write(0, SRAM_CTRL_CTRL_INIT_BIT, true);
+  mmio_region_write32(sram_ctrl->base_addr, SRAM_CTRL_CTRL_REG_OFFSET, reg);
+
+  // Wait for memory to be overwritten with pseudo random data.
+  do {
+    status = sram_ctrl_get_status(sram_ctrl);
+  } while (!(status & kDifSramCtrlStatusInitDone));
+
+  // Check for the errors during memory overwriting.
+  if (status & kDifSramCtrlStatusInitErr) {
```
[the remark this answers is not in the document]
Comment (silvestrst) [reply sw/device/lib/dif/dif_sram_ctrl.c]: Done.

File: sw/device/lib/dif/dif_sram_ctrl.c [hunk 3]
```diff
@@ -64,6 +64,39 @@ static uint32_t sram_ctrl_get_status(const dif_sram_ctrl_t *sram_ctrl) {
...
+dif_result_t dif_sram_ctrl_scramble(const dif_sram_ctrl_t *sram_ctrl) {
+  if (sram_ctrl == NULL) {
+    return kDifBadArg;
+  }
+
+  if (sram_ctrl_locked_ctrl(sram_ctrl)) {
+    return kDifLocked;
+  }
+
+  // Scramble memory addresses.
```
Comment (msfschaffner) [inline sw/device/lib/dif/dif_sram_ctrl.c:76]: This comment and the one below may be misleading, since:
1) both the payload of a 32bit as well as its address will be scrambled
2) the memory is technically always scrambled. by writing to `SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT` we are just requesting a new key to parameterize the scrambling functions for data and address. the block has a default scrambling key that it resets to after reset.

Hence, I would recommend to rephrase this to something like: `// Issue request for new scrambling key`

File: sw/device/lib/dif/dif_sram_ctrl.c [hunk 4]
```diff
@@ -64,6 +64,39 @@ static uint32_t sram_ctrl_get_status(const dif_sram_ctrl_t *sram_ctrl) {
...
+  uint32_t reg =
+      bitfield_bit32_write(0, SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT, true);
+  mmio_region_write32(sram_ctrl->base_addr, SRAM_CTRL_CTRL_REG_OFFSET, reg);
+
+  // Wait for address scrambling to finish.
```
Comment (msfschaffner) [inline sw/device/lib/dif/dif_sram_ctrl.c:81]: Same comment as above.
Maybe rephrase to: `// Wait until the scrambling key has been updated`

File: sw/device/lib/dif/dif_sram_ctrl_unittest.cc [hunk 5]
```diff
@@ -23,6 +23,58 @@ class SramCtrlTest : public Test, public MmioTest {
...
+TEST_F(Scramble, Failure) {
+  EXPECT_READ32(SRAM_CTRL_CTRL_REGWEN_REG_OFFSET, 1);
+
+  // Trigger address scrambling.
+  EXPECT_WRITE32(SRAM_CTRL_CTRL_REG_OFFSET,
+                 {{SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT, true},
+                  {SRAM_CTRL_CTRL_INIT_BIT, false}});
+  EXPECT_READ32(SRAM_CTRL_STATUS_REG_OFFSET, kDifSramCtrlStatusScrKeyValid);
+
+  // Trigger data scrambling.
```
Comment (msfschaffner) [inline sw/device/lib/dif/dif_sram_ctrl_unittest.cc:46]: Same comment for the two comments above.

File: sw/device/lib/dif/dif_sram_ctrl_unittest.cc [hunk 6]
```diff
@@ -23,6 +23,58 @@ class SramCtrlTest : public Test, public MmioTest {
...
+  EXPECT_READ32(SRAM_CTRL_STATUS_REG_OFFSET,
+                std::numeric_limits<uint32_t>::max());
+
+  EXPECT_EQ(dif_sram_ctrl_scramble(&sram_ctrl_), kDifError);
+}
+
+TEST_F(Scramble, Success) {
+  EXPECT_READ32(SRAM_CTRL_CTRL_REGWEN_REG_OFFSET, 1);
+
+  // Trigger address scrambling, and emulate three iteration status read loop.
```
Comment (msfschaffner) [inline sw/device/lib/dif/dif_sram_ctrl_unittest.cc:59]: Same comment as above.

File: sw/device/lib/dif/dif_sram_ctrl_unittest.cc [hunk 7]
```diff
@@ -23,6 +23,58 @@ class SramCtrlTest : public Test, public MmioTest {
...
+  EXPECT_WRITE32(SRAM_CTRL_CTRL_REG_OFFSET,
+                 {{SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT, true},
+                  {SRAM_CTRL_CTRL_INIT_BIT, false}});
+  EXPECT_READ32(SRAM_CTRL_STATUS_REG_OFFSET, 0);
+  EXPECT_READ32(SRAM_CTRL_STATUS_REG_OFFSET, 0);
+  EXPECT_READ32(SRAM_CTRL_STATUS_REG_OFFSET, kDifSramCtrlStatusScrKeyValid);
+
+  // Trigger data scrambling, and emulate three iteration status read loop.
```
Comment (msfschaffner) [inline sw/device/lib/dif/dif_sram_ctrl_unittest.cc:67]: This does not trigger any scrambling. Rather, it triggers the intialization feature, which writes pseudorandom data to the memory.

**Review by msfschaffner — COMMENTED** [review]
Thanks for the updates. 

I mainly have a few comments regarding terminology.

File: sw/device/lib/dif/dif_sram_ctrl.c (hunk 3 above)
Comment (silvestrst) [reply sw/device/lib/dif/dif_sram_ctrl.c]: Done.

File: sw/device/lib/dif/dif_sram_ctrl.c (hunk 4 above)
Comment (silvestrst) [reply sw/device/lib/dif/dif_sram_ctrl.c]: Done.

File: sw/device/lib/dif/dif_sram_ctrl_unittest.cc (hunk 5 above)
Comment (silvestrst) [reply sw/device/lib/dif/dif_sram_ctrl_unittest.cc]: Done.

File: sw/device/lib/dif/dif_sram_ctrl_unittest.cc (hunk 6 above)
Comment (silvestrst) [reply sw/device/lib/dif/dif_sram_ctrl_unittest.cc]: Done.

File: sw/device/lib/dif/dif_sram_ctrl_unittest.cc (hunk 7 above)
Comment (silvestrst) [reply sw/device/lib/dif/dif_sram_ctrl_unittest.cc]: Done.

**Review by silvestrst — COMMENTED** [review]
Thank you @Michal-Sch , I have updated the PR.

**Comment by Michal-Sch** [comment]
> Thank you @Michal-Sch , I have updated the PR.

@silvestrst I think you intended to thank someone else ... :-)

**Comment by silvestrst** [comment]
> > Thank you @Michal-Sch , I have updated the PR.
> 
> @silvestrst I think you intended to thank someone else ... :-)
Thank you as well for pointing it out :)

**Review by mcy — APPROVED** [review]

**Review by msfschaffner — APPROVED** [review]
Thanks for updating the comments!

**Review by doreis-lowrisc — APPROVED** [review]
